### ISSUE TITLE
IAPClient.EnvReady crash

### ISSUE THREAD

**Issue by manelizzard** [EvilMindDevs/hms-unity-plugin#29]
**Environment**:

- Unity 2019.3.0f6
- `master` branch of this repo
- Using Unity IAP with `HuaweiStore` 
- Built with IL2CPP

**Issue**:

IAP does not get fully/properly initialized.

**Description**:

While trying to make IAP work, I get the following error when I invoke `Iap.GetIapClient().envReady`:

```
06-17 10:17:51.963 27735 27735 I HMSSDK_IsEnvReadyTaskApiCall: onResult, success
06-17 10:17:52.011 27735 27735 E Unity   : MissingMethodException: Constructor on type 'HuaweiMobileServices.IAP.EnvReadyResult' not found.
06-17 10:17:52.011 27735 27735 E Unity   :   at System.RuntimeType.CreateInstanceImpl (System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes, System.Threading.StackCrawlMark& stackMark) [0x00000] in <00000000000000000000000000000000>:0 
06-17 10:17:52.011 27735 27735 E Unity   :   at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes) [0x00000] in <00000000000000000000000000000000>:0 
06-17 10:17:52.011 27735 27735 E Unity   :   at System.Activator.CreateInstance (System.Type type, System.Object[] args) [0x00000] in <00000000000000000000000000000000>:0 
06-17 10:17:52.011 27735 27735 E Unity   :   at HuaweiMobileServices.Utils.AndroidJavaObjectExtensions.AsWrapper[T] (UnityEngine.AndroidJavaObject javaObject) [0x00000] in <00000000000000000000000000000000>:0 
```

It cannot find `HuaweiMobileServices.IAP.EnvReadyResult` constructor, but seems all the previous process is completed successfully. 

Because this initialization fails, `IAPClient` does not get fully initialized, and `AddOnSuccessListener` is never called.

**Comment by manelizzard** [comment]
**Second test**:

If I try to request ProductInfo after a prudent time (ignoring `EnvReady` crash), I do get the following error:

```
MissingMethodException: Constructor on type 'HuaweiMobileServices.IAP.ProductInfoResult' not found
```

Seems `HuaweiMobileServices.IAP` namespace or library is not properly included within the project.

**Comment by m0skit0** [comment]
Thanks! This looks like a bug on the SDK, creating issue there: https://github.com/m0skit0/hms-sdk-unity/issues/47

**Comment by m0skit0** [comment]
> Seems HuaweiMobileServices.IAP namespace or library is not properly included within the project.

Code in the SDK is correct. Namespace is accessible from the project, otherwise it wouldn't compile.  There are no other ways to include it. Can you specify which version of the plugin are you using? Are you using the repo code or the packaged plugin?

**Comment by manelizzard** [comment]
I'm using the `master` branch of this repo. As you said, it's probably an SDK issue (https://github.com/m0skit0/hms-sdk-unity/issues/47).

I suspect the wrapper is correct, but somehow the `HuaweiMobileServices.IAP.EnvReadyResult` class is not available in the Android SDK (aar), so when the JNI tries to build that object on the Java side, it crashes.

**Comment by m0skit0** [comment]
I can confirm it's not an SDK issue, the `EnvReadyResult` class in C# has the correct constructor expected by AsWrapper(). The `EnvReadyResult` Java class is available on the Huawei AARs. In any case the wrapper never constructs the object itself, it just wraps the Java object returned from the Huawei AAR. The problem as far as I understand from the stacktrace is that when the Java EnvReadyResult is returned, the AsWrapper() method cannot find the correct constructor for the C# EnvReadyResult wrapper. However the constructor is there, so the error doesn't really makes sense unless you have a different/outdated DLL. 

Since I cannot reproduce this error, to be able to advance in this problem:
- Is there any way you can make a simple demo that can reproduce this error and share it with me?
- Can you build the DLL yourself from the SDK repo and use that one instead?

**Comment by manelizzard** [comment]
Hi,

This repository, when I build and Android APK and install it on my device, fires the above exception. IAP never gets fully initialized. I also tried building the DLL on my own with the SDK repo, but the result is the same.

You say:

> The EnvReadyResult Java class is available on the Huawei AARs.

But if I look in the Android doc for `com.huawei.hms.iap`, the class name is `IsEnvReadyResult` (https://developer.huawei.com/consumer/en/doc/development/HMS-References/iap-IsEnvReadyResult-v4). Seeing the same class name when decompiling the AAR:

<img width="1067" alt="isenvresultready" src="https://user-images.githubusercontent.com/895589/85011855-6ead8b80-b162-11ea-9b67-dc4345552444.png">

So:

- May this class name difference be the cause of not finding the constructor?
- Does the IAP Demo work for you within this repo: build android APK, execute and do a test purchase?

Thank you!

**Comment by m0skit0** [comment]
Yes, sorry, my mistake, the Java class is `IsEnvReadyResult `. In any case the Java class name is not relevant; the error is about the C# constructor, not the Java one. We're not instantiating that Java class anywhere in the SDK, just wrapping the instance we're receiving from the HMS SDK in a C# class. This C# class we do instantiate through reflection using the `AsWrapper()` method:

```
public static T AsWrapper<T>(this AndroidJavaObject javaObject) where T : JavaObjectWrapper =>
            Activator.CreateInstance(typeof(T), javaObject) as T;
```

You can see this method called in the stacktrace of the error (last line). So what this is doing is instantiating C# `EnvReadyResult` with a constructor that has single `AndroidJavaObject` argument. So `EnvReadyResult` should have a constructor that takes `AndroidJavaObject` as a parameter. If you check `EnvReadyResult` you can see it does have that constructor

```
public EnvReadyResult(AndroidJavaObject javaObject) : base(javaObject) { }
```

Now the question remains: why it is complaining about this constructor does not exist?

**Comment by manelizzard** [comment]
Indeed. Totally agree. I can see all constructors and classes.

I'm sill investigating trying to find an answer to that question...

**Comment by m0skit0** [comment]
I think I found the problem. Will fix in the SDK directly.

https://github.com/m0skit0/hms-sdk-unity/issues/49

**Comment by manelizzard** [comment]
Good catch! Will follow that thread. Thank you!

**Comment by m0skit0** [comment]
Fixed in commit 6c21d0193bb346865dbbd99e081bea4d9d23d163
